### PR TITLE
167377549 Create user notification system

### PR DESCRIPTION
**Pull request by Merkll**
#### What does this PR do?
This pull request creates an email and in-app notification system for user activities

#### Description of Task to be completed
- create notification for user-related activities
- schedules notification for batch email sending
- implement batched email notification
- implement in-app notification using socket.io

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#167377549](https://pivotaltracker.com/story/show/167377549)

#### Screenshots (if appropriate)
Screenshot 2019-08-07 at 4 32 00 AM

#### Questions:
N/A

### CODE REVIEW

File: src/services/notification.js [hunk 1]
```diff
@@ -38,6 +43,20 @@ class Notification extends EventEmitter {
...
 
+  /**
+   * Initialises the event
+   *
+   * @param {string} type - the type of notification
+   * @param {Function} notificationHandler - function that handles defined notification
+   *
+   * @return {Notification} -instance of notification
+   */
+  addNotification(type, notificationHandler) {
```
Comment (ibkadeeko) [inline src/services/notification.js:60]: Great Job. Please go through your Js Docs, I believe we are to have `@functions` stated based on our team convention

**Review by ibkadeeko — CHANGES_REQUESTED** [review]

**Review by kennyedward — APPROVED** [review]
This looks good to me.

**Review by ibkadeeko — APPROVED** [review]

File: src/helpers/models.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * @function
+ *
+ * @param {Function} sourceFunction - function to call as data source
+ * @param {Object} dataToFunction - data to pass to source function as parameter
+ * @param {number} size - size of data to return per page
+ *
+ * @return {Object} - sequelize object for inserted data
+ */
+export const QueryPagination = async (sourceFunction, dataToFunction = {}, size = 50) => {
```
Comment (iverenshaguy) [inline src/helpers/models.js]: This is not a Class or a Model, it's a helper and should have a camelCased name. Please fix all functions in this file

File: src/services/email.js [hunk 3]
```diff
@@ -5,24 +5,28 @@ import Debug from 'debug';
...
  * @function
  *
- * @param {String} type - the type of message which corresponds to the mail template
- * @param {Array} data - can be a list for bulk sending. must contain email parameter
+ * @param {string} type - the type of message which corresponds to the mail template
+ * @param {Array} payload - can be a list for bulk sending. must contain email parameter
+ * @param {string} template - sendgrid template to use
  *
- * @return {promise} - Returns a promise that get resolcved to a success object
+ * @return {Promise} - Returns a promise that get resolved to a success object
```
Comment (iverenshaguy) [inline src/services/email.js:23]: Thanks for correcting

File: src/services/email.js [hunk 4]
```diff
@@ -5,24 +5,28 @@ import Debug from 'debug';
...
  */
-export default async ({ type, payload }) => {
+export default async ({ type, payload, template }) => {
   try {
     sendgrid.setApiKey(SENDGRID_API_KEY);
 
-    const template = templates[type];
+    template = template || templates[type];
 
     if (!template) throw new Error('Email Template not available');
```
Comment (iverenshaguy) [inline src/services/email.js]:
```suggestion
    if (!template) throw new Error('Email template not available');
```

**Review by iverenshaguy — APPROVED** [review]
I see a very very good work here. Well done. Please check my review comments and update your PR so that I can merge it

**Pull request merged**